### PR TITLE
createCustomRoutine and createExtendedRoutine

### PR DESCRIPTION
**Pull request by shapkarin**
issue #56 with createRoutineCreator

### CODE REVIEW

**Comment by shapkarin** [comment]
Forgot, I'll add docs soon too.

File: src/createActionCreator.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+const isFunction = value => typeof value === 'function';
+const getCreatorForType = (type, creator) => {
+  if (!creator) {
+    return creator;
+  }
+  if (isFunction(creator[type])) {
+    return creator[type];
+  }
+  if (isFunction(creator[toCamelCase(type)])) {
```
Comment (afitiskin) [inline src/createActionCreator.js]: Do we really need `toCamelCase` here?

File: src/createRoutine.js [hunk 2]
```diff
@@ -1,35 +1,15 @@
...
-    }
-    return undefined;
-  };
-
-  const createActionCreator = (type) => createAction(`${typePrefix}/${type}`, getCreatorForType(type, payloadCreator), getCreatorForType(type, metaCreator));
-
   return routineStages.reduce(
     (result, stage) => {
-      const actionCreator = createActionCreator(stage);
+      const actionCreator = createActionCreator(typePrefix, stage, payloadCreator, metaCreator);
```
Comment (afitiskin) [inline src/createRoutine.js]: Here you lost the ability to set payload/meta creators for every stage separately (e.g. one payload creator for `request` and another for `success`)

File: src/createRoutine.js [hunk 3]
```diff
@@ -1,35 +1,15 @@
-import { createAction } from 'redux-actions';
 import routineStages from './routineStages';
-
-const isFunction = value => typeof value === 'function';
+import createActionCreator from './createActionCreator';
 
 export default function createRoutine(typePrefix, payloadCreator, metaCreator) {
```
Comment (afitiskin) [inline src/createRoutine.js]: In my example `createRoutine` was build on top of `createRoutineCreator` to reduce code duplicates

File: src/createRoutineCreator.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
+import toCamelCase from 'camelcase';
+
+import createRoutine from './createRoutine';
+import createActionCreator from './createActionCreator';
+
+export default function createRoutineCreator(stages = [], payloadCreator, metaCreator){
+  const stagesArray = typeof stages === 'string' ? [stages] : stages;
```
Comment (afitiskin) [inline src/createRoutineCreator.js]: `stagesArray = [].concat(stages)`;

File: src/createRoutineCreator.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+  return function (prefix) {
+    const routine = createRoutine(prefix, payloadCreator, metaCreator);
```
Comment (afitiskin) [inline src/createRoutineCreator.js]: [I prefer not to extend routines but build it by using list of stages](https://github.com/afitiskin/redux-saga-routines/issues/56#issuecomment-501991069)

File: test/createRoutineCreator.js [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+
+
+describe('createRoutineCreator', () => {
+  it('should be a function', () => {
+    expect(createRoutineCreator).to.be.a('function');
+  });
+
+  it('should create routine with empty stages argument', () => {
+    const createExtendedRoutine = createRoutineCreator();
+    const routine = createExtendedRoutine(PREFIX);
```
Comment (afitiskin) [inline test/createRoutineCreator.js]: See no reason to use `createRoutineCreator()` without passing params. Error should be thrown

File: test/createRoutineCreator.js [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+
+    expect(routine.fulfill).to.be.a('function');
+    expect(routine.FULFILL).to.equal(FULFILL);
+    expect(routine.fulfill.toString()).to.equal(FULFILL);
+    expect(routine.fulfill(payload)).to.deep.equal(fulfillAction);
+  });
+
+
+  it('should create an extened routine', () => {
+    const createExtendedRoutine = createRoutineCreator(['SOME_LONG_TYPE', 'OTHER']);
```
Comment (afitiskin) [inline test/createRoutineCreator.js]: I still do not understand why and for what do you need a routine with 7 stages?

File: test/createRoutineCreator.js [hunk 8]
```diff
@@ -0,0 +1,334 @@
...
+    expect(routine.someLongType(payload)).to.deep.equal(someAction);
+
+    expect(routine.other).to.be.a('function');
+    expect(routine.OTHER).to.equal(OTHER);
+    expect(routine.other.toString()).to.equal(OTHER);
+    expect(routine.other(payload)).to.deep.equal(otherAction);
+  });
+
+  it('should create extened routine if types arg is a string', () => {
+    const createExtendedRoutine = createRoutineCreator('SOME_LONG_TYPE');
```
Comment (afitiskin) [inline test/createRoutineCreator.js]: Any examples of stage longer than one word? IMO stage is a short and informative verb: show, hide, open, close, toggle, trigger, request, remove, populate, truncate, ...

File: src/createRoutine.js (hunk 2 above)
Comment (shapkarin) [reply src/createRoutine.js]: I test payload separately creators https://github.com/afitiskin/redux-saga-routines/pull/59/commits/bae356a951a40f41b1acf283050c8e3711071070

File: src/createActionCreator.js (hunk 1 above)
Comment (shapkarin) [reply src/createActionCreator.js]: to I need that match long custom stages like `SOME_LOMNG_STAGE` with key `someSongStage` to add `payloadCreator` or `metaCreator`

File: test/createRoutineCreator.js (hunk 8 above)
Comment (shapkarin) [reply test/createRoutineCreator.js]: `TOGGLE_PROJECT_INFO` https://github.com/shapkarin/shapkarin.me/tree/master/src/Pages/Projects

sure, I can name that `TOGGLE` but anyway that `stages` are goes direct to `redux types` and types can be any length.

I imagine that PR feature like I nice way to have all types and actions about some "enity" not only in one namespace but also in one "enity" like yours `routine`. Seems that it makes code more consistent.

**Comment by shapkarin** [comment]
>  that can be used to create your own customCreateRoutine function, with completely different set of stages and then, use it the same way you are using the standard createRoutine function.

But what if I want to create a custom routine with all that default stages like `trigger` and so on? With  `CombineRoutines()` two routines next?

**Comment by shapkarin** [comment]
`CombineRoutines` also is not a good idea because routines can have different prefix. 

P.S. At my comment to `test/createRoutineCreator.js` conversation you can find some more info about the feature I imagine.

**Comment by shapkarin** [comment]
But I like the idea to also have a function that can create routine without default stages.

File: test/createRoutineCreator.js (hunk 7 above)
Comment (shapkarin) [reply test/createRoutineCreator.js]: I wrote about that at other conversation

**Comment by shapkarin** [comment]
> I imagine that PR feature like I nice way to have all types and actions about some "enity" not only in one namespace but also in one "enity" like yours routine. Seems that it makes code more consistent.

Stages can be any length and describe anything that can describe action `type`. It's also great to have function like `createCustumRoutine()` but I also need something that create routine with default stages and also my own. 

I hope that all is clear now. And sure, I'll try to avoid any breaking changes especially if it's about project's public API.

**Comment by shapkarin** [comment]
> I imagine that PR feature like I nice way to have all types and actions about some "enity" not only in one namespace but also in one "enity" like yours routine. Seems that it makes code more consistent.

Stage can has any length, count of stages also can be any, and they describe anything that can describe `actions` and `types`. It's also great to have a function like `createCustumRoutine()` with completely different set of stages but I also need something that create routine with default stages and also my own. 

I hope that all is clear now. And sure, I'll try to avoid any breaking changes especially if it's about project's public API.

File: test/createRoutineCreator.js (hunk 6 above)
Comment (shapkarin) [reply test/createRoutineCreator.js]: ok

**Comment by shapkarin** [comment]
I got what you comment about `createRoutineCreator` low-level usage

didn't pay attention to that line:
```js
export default createRoutineCreator(defaultRoutineStages);
```

I'll change my PR soon, but please have a time to thing about what I described before, about new features `createCustumRoutine()` and `createExtendedRoutine()` and what stages means in that context.

File: src/createActionCreator.js (hunk 1 above)
Comment (shapkarin) [reply src/createActionCreator.js]: I can also use 
```
const toCamelCase = string => string.toLowerCase().replace(/[_.\- ]+(\w|$)/g, (_, p1) => p1.toUpperCase());
```
but it feels wrong to copy that from `camelcase` package

Comment (shapkarin) [reply src/createActionCreator.js]: maybe I can just use something like:
```
const toCamelCase = string => string.toLowerCase().replace(/_+(\w|$)/g, (_, p1) => p1.toUpperCase());
```

**Comment by shapkarin** [comment]
@afitiskin just what to check if I got all right and remove breaking changes. If it's OK next I'll add docs.

File: src/createRoutine.js (hunk 3 above)
Comment (shapkarin) [reply src/createRoutine.js]: I don't know why it's now marked as `outdated` https://github.com/afitiskin/redux-saga-routines/pull/59/files#diff-9212afb16a80781533a3d2d29a222017

Comment (shapkarin) [reply src/createRoutine.js]: I fixed that

File: README.md [hunk 9]
```diff
@@ -441,6 +442,41 @@ function* sendFormDataToServer(formData) {
...
+
+projects.TOGGLE === 'projects/TOGGLE';
+projects.toggle({ id: 42 }) === { type: "projects/TOGGLE", payload: { id: 42 } };
+```
+
+You can pass an array or stages:
+```javascript
+import { createExtendedRoutine } from 'redux-saga-routines';
+
+const other = createExtendedRoutine('other', ['OPEN', 'CLOSE']);
```
Comment (shapkarin) [inline README.md]: I'll try to add better docs and examples for `createExtendedRoutine` and `createCustomRoutine`

File: README.md [hunk 10]
```diff
@@ -441,6 +442,41 @@ function* sendFormDataToServer(formData) {
...
+
+other.OPEN === 'other/OPEN';
+other.open(42) === { type: "other/OPEN", payload: 42 }
+
+other.CLOSE === 'other/CLOSE';
+other.close(42) === { type: "other/CLOSE", payload: 42 }
```
[the remark this answers is not in the document]
Comment (shapkarin) [reply README.md]: or 
```js
expect(other.close(42)).to.deep.equal({ type: "other/CLOSE", payload: 42 });
```

**Comment by afitiskin** [comment]
Sorry for the delay I was on vacation. I will review the code, apply some changes and release new version soon

**Comment by shapkarin** [comment]
Then, I'll try find time to update docs too

File: README.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (shapkarin) [reply README.md]: @afitiskin what if I update all docs with `console.log` ?

File: README.md [hunk 11]
```diff
@@ -441,6 +442,56 @@ function* sendFormDataToServer(formData) {
 }
 ```
 
+### You also can create routie with default stages and yours custom with `createExtendedRoutine`:
```
Comment (shapkarin) [inline README.md]: Not sure if that part must be placed at the end of the docs

File: README.md [hunk 12]
```diff
@@ -441,6 +442,56 @@ function* sendFormDataToServer(formData) {
...
+other.CLOSE === 'other/CLOSE';
+console.log(other.close(42)) // { type: "other/CLOSE", payload: 42 }
+```
+
+And also you can add cusom payload and meta creators with the same API as `createRoutine`:
+```javascript
+import { createExtendedRoutine } from 'redux-saga-routines';
+
+const customPayloadMeta = createExtendedRoutine('payload/meta', 'MILTIPLIED_PAYLOAD',
+  { multipliedPayload: (payload) => payload * 2 },
```
Comment (shapkarin) [inline README.md]: Add separated note about function camelCased names

File: README.md (hunk 11 above)
Comment (shapkarin) [reply README.md]: I mean all about `createExtendedRoutine` and `createCustomRoutine`

File: README.md [hunk 13]
```diff
@@ -441,6 +442,59 @@ function* sendFormDataToServer(formData) {
...
+const steps = createCustomRoutine('steps', ['NEXT', 'PREVIOUS', 'TO']);
+
+steps.NEXT === 'steps/NEXT';
+console.log(steps.next()); // { type: "steps/NEXT" }
+
+steps.PREVIOUS === 'steps/PREVIOUS';
+console.log(steps.previous()); // { type: "steps/PREVIOUS" }
+
+steps.TO === 'steps/TO';
+console.log(steps.to(3)); // { type: "steps/PREVIOUS", payload: 3 }
```
Comment (shapkarin) [inline README.md]: Maybe better is rename to ‘GO_TO’ and next ‘steps.goTo(3)’

**Comment by shapkarin** [comment]
@afitiskin I update docs and squash commits.

File: README.md (hunk 11 above)
Comment (shapkarin) [reply README.md]: I hope I choose the right new place

File: test/createRoutineCreator.js (hunk 8 above)
Comment (shapkarin) [reply test/createRoutineCreator.js]: I added an example with [steps routine](https://github.com/afitiskin/redux-saga-routines/pull/59/files#diff-04c6e90faac2675aa89e2176d2eec7d8R99)

**Comment by darewreck54** [comment]
@afitiskin When will this be released?

**Comment by shapkarin** [comment]
Hi, @darewreck54, as you can see it's merged but if you are looking for the same API that I mention after this discussion you probably will find it in [that package](https://github.com/shapkarin/extend-saga-routines) soon.

**Comment by afitiskin** [comment]
I release `3.2.0` today, as I mentioned in #56, I decided to exclude `createExtendedRoutine` and `createCustomRoutine` from the package, but it's just a few lines of code, that can be done in your app:

```javascript
import { creteRoutineCreator, routineStages } from 'redux-saga-routines';
export const createCustomRoutine = (typePrefix, stages, payloadCreator, metaCreator) => creteRoutineCreator(stages)(typePrefix, payloadCreator, metaCreator);
export const createExtendedRoutine = (typePrefix, stages, payloadCreator, metaCreator) => creteRoutineCreator(routineStages.concat(stages))(typePrefix, payloadCreator, metaCreator);
```

**Comment by shapkarin** [comment]
@afitiskin, sure I saw that comment at issue, np, I plan to implement `createCustomRoutine` and `createExtendedRoutine` as part of `extend-saga-routines`. Thanks that new version allows me to make that easily.

**Comment by shapkarin** [comment]
If anyone wonder about `createExtendedRoutine` and `createCustomRoutine`, they can use that package https://github.com/shapkarin/extend-saga-routines
It needs `redux-saga-routines` version starts from `3.2.0` installed in yours projects.
Also with that package you can extend any already created routine with `extendRoutine` function.

**Pull request merged**